File: Viscoacoustic Modeling/11. Python_GPU_Numba_CUPY/AC2D.py
"""
AC2D object

This script defines a AC2D class for solving the acoustic wave equation in 2D.

@author: Stefan Catalin Craciun
"""

# Import Libraries 
# ----------------
import matplotlib.pyplot as plt
import numpy as np

from numba import jit, float64, int64
from numba.experimental import jitclass
from numba import prange
import numba
import cupy as cp
import numba.cuda as cuda
import timeit

# Import Classes  
# ----------------
from Differentiator import * 
from Model import * 
from Src import * 
from Rec import * 

                            

spec = [
    ('p', float64[:,:]),
    ('vx', float64[:,:]),
    ('vy', float64[:,:]),
    ('exx', float64[:,:]),
    ('eyy', float64[:,:]),
    ('gammax', float64[:,:]),
    ('gammay', float64[:,:]),
    ('thetax', float64[:,:]),
    ('thetay', float64[:,:]),
    ('ts', int64),
]
# ----------------
# AC2D Class
# ----------------
#@jitclass(spec)
class Ac2d:
    def __init__(self, Model):
        """
        Initializes an Ac2d object with the given model object.

        Parameters
        ----------
        Model : Model object
                The model object containing the simulation parameters.
        """

        self.p = np.zeros((Model.Nx, Model.Ny))                     # Stress
        self.vx = np.zeros((Model.Nx, Model.Ny))                    # Particle velocity in x-direction
        self.vy = np.zeros((Model.Nx, Model.Ny))                    # Particle velocity in y-direction
        self.exx = np.zeros((Model.Nx, Model.Ny))                   # Time derivative of Strain in x-direction
        self.eyy = np.zeros((Model.Nx, Model.Ny))                   # Time derivative of Strain in y-direction
        
        self.gammax = np.zeros((Model.Nx, Model.Ny), dtype=float)   
        self.gammay = np.zeros((Model.Nx, Model.Ny), dtype=float)
        self.thetax = np.zeros((Model.Nx, Model.Ny), dtype=float)
        self.thetay = np.zeros((Model.Nx, Model.Ny), dtype=float)
        
        self.ts = 0
    
    def Ac2dSolve(Ac2d, Model, Src, Rec, nt, l):
        """
        Ac2dSolve computes the solution of the acoustic wave equation.
        The acoustic equation of motion are integrated using Virieux's (1986) stress-velocity scheme.
        (See the notes.tex file in the Doc directory).
        
            vx(t+dt)   = dt/rhox d^+x[ sigma(t)] + dt fx + vx(t)
                       + thetax D[1/rhox]
            vy(t+dt)   = dt/rhoy d^+y sigma(t) + dt fy(t) + vy(t)
                       + thetay D[1/rhoy]

            dp/dt(t+dt) = dt Kappa[d^-x dexx/dt + d-y deyy/dt + dt dq/dt(t) 
                        + dt [gammax Dkappa + gammay Dkappa]
                        + p(t)

            dexx/dt     =  d^-_x v_x 
            deyy/dt     =  d^-_z v_y 

            gammax(t+dt) = alpha1x gammax(t) + alpha2x dexx/dt 
            gammay(t+dt) = alpha1y gammay(t) + alpha2y deyy/dt 

            thetax(t+dt) = eta1x thetax(t) + eta2x d^+x p
            thetay(t+dt) = eta1y thetay(t) + eta2y d^+y p
            
            Parameters:
                Ac2d : Solver object
                Model: Model object
                Src  : Source object
                Rec  : Receiver object
                nt   : Number of timesteps to do starting with current step 
                l    : The differentiator operator length
        
        """
        
        Diff = Differentiator(l)                            # Create differentiator object
        oldperc = 0.0                                       # Old percentage for printing progress
        ns = Ac2d.ts                                        # Start timestep
        ne = ns + nt                                        # Stop timestep

        sx = int(round(Src.Sx[0]))
        sy = int(round(Src.Sy[0]))

        # Define the CuPy arrays
        p_gpu = cp.asarray(Ac2d.p)
        vx_gpu = cp.asarray(Ac2d.vx)
        vy_gpu = cp.asarray(Ac2d.vy)
        exx_gpu = cp.asarray(Ac2d.exx)
        eyy_gpu = cp.asarray(Ac2d.eyy)
        gammax_gpu = cp.asarray(Ac2d.gammax)
        gammay_gpu = cp.asarray(Ac2d.gammay)
        thetax_gpu = cp.asarray(Ac2d.thetax)
        thetay_gpu = cp.asarray(Ac2d.thetay)
        rho_gpu = cp.asarray(Model.Rho)
        kappa_gpu = cp.asarray(Model.Kappa)
        drhox_gpu = cp.asarray(Model.Drhox)
        drhoy_gpu = cp.asarray(Model.Drhoy)
        dkappax_gpu = cp.asarray(Model.Dkappax)
        dkappay_gpu = cp.asarray(Model.Dkappay)
        eta1x_gpu = cp.asarray(Model.Eta1x)
        eta1y_gpu = cp.asarray(Model.Eta1y)
        eta2x_gpu = cp.asarray(Model.Eta2x)
        eta2y_gpu = cp.asarray(Model.Eta2y)
        alpha1x_gpu = cp.asarray(Model.Alpha1x)
        alpha1y_gpu = cp.asarray(Model.Alpha1y)
        alpha2x_gpu = cp.asarray(Model.Alpha2x)
        alpha2y_gpu = cp.asarray(Model.Alpha2y)
        src_gpu = cp.asarray(Src.Src)
        w_gpu = cp.asarray(Diff.w)


        # Set up the grid and block dimensions
        #threads_per_block = 1024
        #blocks_per_grid = (x.size + (threads_per_block - 1)) // threads_per_block

        # Calculate the grid and block dimensions
        block_dim = (16, 16)
        grid_dim = (int(np.ceil(Model.Nx / block_dim[0])), int(np.ceil(Model.Ny / block_dim[1])))

        print("START________________________________")
        cp.cuda.Stream.null.synchronize()
        
        # Loop over timesteps
        for i in range(ns, ne):
            if i==1:
                start = timeit.default_timer()
            # Compute spatial derivative of stress
            # Use exx and eyy as temp storage
            DiffDxplus_kernel[grid_dim, block_dim](exx_gpu, p_gpu, w_gpu, Model.Dx, Model.Nx, Model.Ny, l)     # Forward differentiation x-axis
            Ac2dvx_kernel[grid_dim, block_dim](vx_gpu, exx_gpu, thetax_gpu, rho_gpu, drhox_gpu, eta1x_gpu, eta2x_gpu, Model.Dt, Model.Nx, Model.Ny)                             # Compute vx

            DiffDyplus_kernel[grid_dim, block_dim](eyy_gpu, p_gpu, w_gpu, Model.Dx, Model.Nx, Model.Ny, l)     # Forward differentiation y-axis
            Ac2dvy_kernel[grid_dim, block_dim](vy_gpu, eyy_gpu, thetay_gpu, rho_gpu, drhoy_gpu, eta1y_gpu, eta2y_gpu, Model.Dt, Model.Nx, Model.Ny)                              # Compute vy

            # Compute time derivative of strains
            #Diff.DiffDxminus(Ac2d.vx, Ac2d.exx, Model.Dx)   # Compute exx
            DiffDxminus_kernel[grid_dim, block_dim](exx_gpu, vx_gpu, w_gpu, Model.Dx, Model.Nx, Model.Ny, l)

            #Diff.DiffDyminus(Ac2d.vy, Ac2d.eyy, Model.Dx)   # Compute eyy
            DiffDyminus_kernel[grid_dim, block_dim](eyy_gpu, vy_gpu, w_gpu, Model.Dx, Model.Nx, Model.Ny, l)

            # Update stress
            #Ac2d.Ac2dstress(Model)
            Ac2dstress_kernel[grid_dim, block_dim](p_gpu, exx_gpu, eyy_gpu, gammax_gpu, gammay_gpu, kappa_gpu, dkappax_gpu, dkappay_gpu, alpha1x_gpu, alpha1y_gpu, alpha2x_gpu, alpha2y_gpu, Model.Dt, Model.Nx, Model.Ny)


            # Add source
            update_pressure_kernel(p_gpu, rho_gpu, Src.Src[i], Model.Dt, Model.Dx, sx, sy)
            #Ac2d.p[sx, sy] += Model.Dt * (Src.Src[i] / (Model.Dx * Model.Dx * Model.Rho[sx,sy]))    


            # Print progress
            perc = 1000.0 * (float(i) / float(ne - ns - 1)) # Current Percentage of completion
            if perc - oldperc >= 10.0:
                iperc = int(perc) // 10                     
                if iperc % 10 == 0:
                    print(iperc)                            # Percentage finished
                oldperc = perc


            Ac2d.p = cp.asnumpy(p_gpu)
            # Record Wavefield
            Rec.rec_wavefield(i, Ac2d.p)
            # Record Seismogram
            Rec.rec_seismogram(i, Ac2d.p)
            # Record 1 Trace
            Rec.rec_trace(i, Ac2d.p)
            
        cp.cuda.Stream.null.synchronize()
        end = timeit.default_timer() 
        
        print("Execution Time: ", (end - start))
        #print(f"Size of one element: {Rec.wavefield.itemsize} bytes")
        #print(f"Size of the whole array: {Rec.wavefield.nbytes} bytes")
        cp.cuda.MemoryPool().free_all_blocks()
        return "OK"

        
def update_pressure_kernel(p_gpu, rho_gpu, src_val, Dt, Dx, sx, sy):
    p_gpu[sx, sy] += Dt * (src_val / (Dx * Dx * rho_gpu[sx, sy]))

@cuda.jit
def Ac2dvx_kernel(vx_gpu, exx_gpu, thetax_gpu, rho_gpu, drhox_gpu, eta1x_gpu, eta2x_gpu, dt, nx, ny):
    x, y = cuda.grid(2)

    if x < nx and y < ny:
        vx_gpu[x, y] = dt * (1.0 / rho_gpu[x, y]) * exx_gpu[x, y] + vx_gpu[x, y] + dt * thetax_gpu[x, y] * drhox_gpu[x, y]
        thetax_gpu[x, y] = eta1x_gpu[x, y] * thetax_gpu[x, y] + eta2x_gpu[x, y] * exx_gpu[x, y]

@cuda.jit
def Ac2dvy_kernel(vy_gpu, eyy_gpu, thetay_gpu, rho_gpu, drhoy_gpu, eta1y_gpu, eta2y_gpu, dt, nx, ny):
    x, y = cuda.grid(2)

    if x < nx and y < ny:
        vy_gpu[x, y] = dt * (1.0 / rho_gpu[x, y]) * eyy_gpu[x, y] + vy_gpu[x, y] + dt * thetay_gpu[x, y] * drhoy_gpu[x, y]
        thetay_gpu[x, y] = eta1y_gpu[x, y] * thetay_gpu[x, y] + eta2y_gpu[x, y] * eyy_gpu[x, y]


@cuda.jit
def Ac2dstress_kernel(p_gpu, exx_gpu, eyy_gpu, gammax_gpu, gammay_gpu, kappa_gpu, dkappax_gpu, dkappay_gpu, alpha1x_gpu, alpha1y_gpu, alpha2x_gpu, alpha2y_gpu, dt, nx, ny):
    x, y = cuda.grid(2)

    if x < nx and y < ny:
        p_gpu[x, y] = dt * kappa_gpu[x, y] * (exx_gpu[x, y] + eyy_gpu[x, y]) + p_gpu[x, y] + dt * (gammax_gpu[x, y] * dkappax_gpu[x, y] + gammay_gpu[x, y] * dkappay_gpu[x, y])
        gammax_gpu[x, y] = alpha1x_gpu[x, y] * gammax_gpu[x, y] + alpha2x_gpu[x, y] * exx_gpu[x, y]
        gammay_gpu[x, y] = alpha1y_gpu[x, y] * gammay_gpu[x, y] + alpha2y_gpu[x, y] * eyy_gpu[x, y]

@cuda.jit
def DiffDyminus_kernel(dA_gpu, A_gpu, w_gpu, dx, nx, ny, l):
    i, j = cuda.grid(2)

    if i < nx and j < ny:
        if j < l:
            sum = 0.0
            for k in range(1, j + 1):
                sum -= w_gpu[k - 1] * A_gpu[i, j - k]
            for k in range(1, l + 1):
                sum += w_gpu[k - 1] * A_gpu[i, j + (k - 1)]
            dA_gpu[i, j] = sum / dx

        elif j >= l and j < ny - l:
            sum = 0.0
            for k in range(1, l + 1):
                sum += w_gpu[k - 1] * (-A_gpu[i, j - k] + A_gpu[i, j + (k - 1)])
            dA_gpu[i, j] = sum / dx

        elif j >= ny - l:
            sum = 0.0
            for k in range(1, l + 1):
                sum -= w_gpu[k - 1] * A_gpu[i, j - k]
            for k in range(1, ny - j + 1):
                sum += w_gpu[k - 1] * A_gpu[i, j + (k - 1)]
            dA_gpu[i, j] = sum / dx

@cuda.jit
def DiffDxminus_kernel(dA_gpu, A_gpu, w_gpu, dx, nx, ny, l):
    i, j = cuda.grid(2)

    if i < nx and j < ny:
        if i < l:
            sum = 0.0
            for k in range(1, i + 1):
                sum -= w_gpu[k - 1] * A_gpu[i - k, j]
            for k in range(1, l + 1):
                sum += w_gpu[k - 1] * A_gpu[i + (k - 1), j]
            dA_gpu[i, j] = sum / dx

        elif i >= l and i < nx - l:
            sum = 0.0
            for k in range(1, l + 1):
                sum += w_gpu[k - 1] * (-A_gpu[i - k, j] + A_gpu[i + (k - 1), j])
            dA_gpu[i, j] = sum / dx

        elif i >= nx - l:
            sum = 0.0
            for k in range(1, l + 1):
                sum -= w_gpu[k - 1] * A_gpu[i - k, j]
            for k in range(1, nx - i + 1):
                sum += w_gpu[k - 1] * A_gpu[i + (k - 1), j]
            dA_gpu[i, j] = sum / dx

@cuda.jit
def DiffDyplus_kernel(dA_gpu, A_gpu, w_gpu, dx, nx, ny, l):
    i, j = cuda.grid(2)

    if i < nx and j < ny:
        if j < l:
            sum = 0.0
            for k in range(1, j + 2):
                sum -= w_gpu[k - 1] * A_gpu[i, j - (k - 1)]
            for k in range(1, l + 1):
                sum += w_gpu[k - 1] * A_gpu[i, j + k]
            dA_gpu[i, j] = sum / dx

        elif j >= l and j < ny - l:
            sum = 0.0
            for k in range(1, l + 1):
                sum += w_gpu[k - 1] * (-A_gpu[i, j - (k - 1)] + A_gpu[i, j + k])
            dA_gpu[i, j] = sum / dx

        elif j >= ny - l:
            sum = 0.0
            for k in range(1, l + 1):
                sum -= w_gpu[k - 1] * A_gpu[i, j - (k - 1)]
            for k in range(1, ny - j):
                sum += w_gpu[k - 1] * A_gpu[i, j + k]
            dA_gpu[i, j] = sum / dx


@cuda.jit
def DiffDxplus_kernel(dA_gpu, A_gpu, w_gpu, dx, nx, ny, l):
    i, j = cuda.grid(2)

    if i < nx and j < ny:

        if i < l:
            sum = 0.0
            for k in range(1, i + 2):
                sum -= w_gpu[k - 1] * A_gpu[i - (k - 1), j]
            for k in range(1, l + 1):
                sum += w_gpu[k - 1] * A_gpu[i + k, j]

            dA_gpu[i, j] = sum / dx

        elif i >= l and i < nx - l:
            sum = 0.0
            for k in range(1, l + 1):
                sum += w_gpu[k - 1] * (-A_gpu[i - (k - 1), j] + A_gpu[i + k, j])
            dA_gpu[i, j] = sum / dx

        elif i >= nx - l:
            sum = 0.0
            for k in range(1, l + 1):
                sum -= w_gpu[k - 1] * A_gpu[i - (k - 1), j]
            for k in range(1, nx - i):
                sum += w_gpu[k - 1] * A_gpu[i + k, j]

            dA_gpu[i, j] = sum / dx

File: Viscoacoustic Modeling/13. Julia_GPU_CUDA/AC2D.jl
"""
AC2D object

This script defines a AC2D mutable struct for solving the acoustic wave equation in 2D.

@author: Stefan Catalin Craciun
"""

using CUDA

#CUDA.versioninfo()


# Import Classes
# ----------------
include("Model.jl")
include("Src.jl")
include("Rec.jl")
include("Differentiator.jl")
# ----------------
# AC2D Type
# ----------------
mutable struct Ac2d
    p::Array{Float64, 2}
    vx::Array{Float64, 2}
    vy::Array{Float64, 2}
    exx::Array{Float64, 2}
    eyy::Array{Float64, 2}
    gammax::Array{Float64, 2}
    gammay::Array{Float64, 2}
    thetax::Array{Float64, 2}
    thetay::Array{Float64, 2}
    ts::Int64
end

function Ac2d(model::Model)
    p = zeros(model.Nx, model.Ny)
    vx = zeros(model.Nx, model.Ny)
    vy = zeros(model.Nx, model.Ny)
    exx = zeros(model.Nx, model.Ny)
    eyy = zeros(model.Nx, model.Ny)
    gammax = zeros(model.Nx, model.Ny)
    gammay = zeros(model.Nx, model.Ny)
    thetax = zeros(model.Nx, model.Ny)
    thetay = zeros(model.Nx, model.Ny)
    ts = 1
    return Ac2d(p, vx, vy, exx, eyy, gammax, gammay, thetax, thetay, ts)
end



function Ac2dSolve(ac2d::Ac2d, model::Model, src::Src, rec::Rec, nt::Int64, l::Int64)
    diff = Differentiator(l)
    oldperc = 0.0
    ns = ac2d.ts
    ne = ns + nt
    sx = round(Int, src.Sx[1])
    sy = round(Int, src.Sy[1])

    # Define the CUDA arrays
    p_gpu = CuArray(ac2d.p)
    vx_gpu = CuArray(ac2d.vx)
    vy_gpu = CuArray(ac2d.vy)
    exx_gpu = CuArray(ac2d.exx)
    eyy_gpu = CuArray(ac2d.eyy)
    gammax_gpu = CuArray(ac2d.gammax)
    gammay_gpu = CuArray(ac2d.gammay)
    thetax_gpu = CuArray(ac2d.thetax)
    thetay_gpu = CuArray(ac2d.thetay)
    rho_gpu = CuArray(model.Rho)
    kappa_gpu = CuArray(model.Kappa)
    drhox_gpu = CuArray(model.Drhox)
    drhoy_gpu = CuArray(model.Drhoy)
    dkappax_gpu = CuArray(model.Dkappax)
    dkappay_gpu = CuArray(model.Dkappay)
    eta1x_gpu = CuArray(model.Eta1x)
    eta1y_gpu = CuArray(model.Eta1y)
    eta2x_gpu = CuArray(model.Eta2x)
    eta2y_gpu = CuArray(model.Eta2y)
    alpha1x_gpu = CuArray(model.Alpha1x)
    alpha1y_gpu = CuArray(model.Alpha1y)
    alpha2x_gpu = CuArray(model.Alpha2x)
    alpha2y_gpu = CuArray(model.Alpha2y)
    src_gpu = CuArray(src.Src)
    w_gpu = CuArray(diff.w)
    
    
    # Determine the block and grid size
    blockDim = (16, 16)
    gridDim = (div(model.Nx + blockDim[1] - 1, blockDim[1]), div(model.Ny + blockDim[2] - 1, blockDim[2]))

    println("START________________________________")
    # Perform necessary operations on the GPU
    for i in ns:ne-1
        #DiffDxplus(diff, ac2d.p, ac2d.exx, model.Dx)
        #DiffDxplus_test(ac2d.p, ac2d.exx, diff.w, model.Dx, model.Nx, model.Ny, l)
        if i == 2
            global t1 = time()
        end
        CUDA.@sync @cuda(
        threads = blockDim,
        blocks = gridDim,
        DiffDxplus_kernel!(exx_gpu, p_gpu, w_gpu, model.Dx, model.Nx, model.Ny, l)
        )

        CUDA.@sync @cuda(
        threads = blockDim,
        blocks = gridDim,
        Ac2dvx_kernel!(vx_gpu, exx_gpu, thetax_gpu, rho_gpu, drhox_gpu, eta1x_gpu, eta2x_gpu, model.Dt, model.Nx, model.Ny)
        )

        #DiffDyplus(diff, ac2d.p, ac2d.eyy, model.Dx)
        CUDA.@sync @cuda(
        threads = blockDim,
        blocks = gridDim,
        DiffDyplus_kernel!(eyy_gpu, p_gpu, w_gpu, model.Dx, model.Nx, model.Ny, l)
        )


        CUDA.@sync @cuda(
        threads = blockDim,
        blocks = gridDim,
        Ac2dvy_kernel!(vy_gpu, eyy_gpu, thetay_gpu, rho_gpu, drhoy_gpu, eta1y_gpu, eta2y_gpu, model.Dt, model.Nx, model.Ny)
        )

        #DiffDxminus(diff, ac2d.vx, ac2d.exx, model.Dx)
        CUDA.@sync @cuda(
        threads = blockDim,
        blocks = gridDim,
        DiffDxminus_kernel!(exx_gpu, vx_gpu, w_gpu, model.Dx, model.Nx, model.Ny, l)
        )

        #DiffDyminus(diff, ac2d.vy, ac2d.eyy, model.Dx)
        CUDA.@sync @cuda(
        threads = blockDim,
        blocks = gridDim,
        DiffDyminus_kernel!(eyy_gpu, vy_gpu, w_gpu, model.Dx, model.Nx, model.Ny, l)
        )

        
        CUDA.@sync @cuda(
        threads = blockDim,
        blocks = gridDim,
        Ac2dstress_kernel!(p_gpu, exx_gpu, eyy_gpu, gammax_gpu, gammay_gpu, kappa_gpu, dkappax_gpu, dkappay_gpu, alpha1x_gpu, alpha1y_gpu, alpha2x_gpu, alpha2y_gpu, model.Dt, model.Nx, model.Ny)
        )

        #Ac2dstress(ac2d, model)
        
     
        CUDA.@sync @cuda update_pressure_kernel!(p_gpu, rho_gpu, src.Src[i], model.Dt, model.Dx, sx, sy)

        #ac2d.p[sx, sy] += model.Dt * (src.Src[i] / (model.Dx * model.Dx * model.Rho[sx, sy]))
        
        
        perc = 1000.0 * (i / (ne - ns - 1))
        if perc - oldperc >= 10.0
            iperc = round(Int, perc) ÷ 10
            if iperc % 10 == 0
                println(iperc)
            end
            oldperc = perc
        end

        ac2d.p = Array(p_gpu)
        rec_wavefield!(rec, i, ac2d.p)
        rec_seismogram!(rec, i, ac2d.p)
        rec_trace!(rec, i, ac2d.p)
        

    end
    t2 = time()
    println("Solver wall clock time: ", t2 - t1)
    return "OK"
end

####### CUDA kernels #######

function Ac2dvx_kernel!(vx_gpu, exx_gpu, thetax_gpu, rho_gpu, drhox_gpu, eta1x_gpu, eta2x_gpu, dt, nx, ny)
    x = threadIdx().x + blockDim().x * (blockIdx().x - 1)
    y = threadIdx().y + blockDim().y * (blockIdx().y - 1)

    if x <= nx && y <= ny
        @inbounds vx_gpu[x, y] = dt * (1.0 / rho_gpu[x, y]) * exx_gpu[x, y] + vx_gpu[x, y] + dt * thetax_gpu[x, y] * drhox_gpu[x, y]
        @inbounds thetax_gpu[x, y] = eta1x_gpu[x, y] * thetax_gpu[x, y] + eta2x_gpu[x, y] * exx_gpu[x, y]
    end
    return
end

function Ac2dvy_kernel!(vy_gpu, eyy_gpu, thetay_gpu, rho_gpu, drhoy_gpu, eta1y_gpu, eta2y_gpu, dt, nx, ny)
    x = threadIdx().x + blockDim().x * (blockIdx().x - 1)
    y = threadIdx().y + blockDim().y * (blockIdx().y - 1)

    if x <= nx && y <= ny
        @inbounds vy_gpu[x, y] = dt * (1.0 / rho_gpu[x, y]) * eyy_gpu[x, y] + vy_gpu[x, y] + dt * thetay_gpu[x, y] * drhoy_gpu[x, y]
        @inbounds thetay_gpu[x, y] = eta1y_gpu[x, y] * thetay_gpu[x, y] + eta2y_gpu[x, y] * eyy_gpu[x, y]
    end
    return
end

function Ac2dstress_kernel!(p_gpu, exx_gpu, eyy_gpu, gammax_gpu, gammay_gpu, kappa_gpu, dkappax_gpu, dkappay_gpu, alpha1x_gpu, alpha1y_gpu, alpha2x_gpu, alpha2y_gpu, dt, nx, ny)
    x = threadIdx().x + blockDim().x * (blockIdx().x - 1)
    y = threadIdx().y + blockDim().y * (blockIdx().y - 1)

    if x <= nx && y <= ny
        @inbounds p_gpu[x, y] = dt * kappa_gpu[x, y] * (exx_gpu[x, y] + eyy_gpu[x, y]) + p_gpu[x, y] + dt * (gammax_gpu[x, y] * dkappax_gpu[x, y] + gammay_gpu[x, y] * dkappay_gpu[x, y])
        @inbounds gammax_gpu[x, y] = alpha1x_gpu[x, y] * gammax_gpu[x, y] + alpha2x_gpu[x, y] * exx_gpu[x, y]
        @inbounds gammay_gpu[x, y] = alpha1y_gpu[x, y] * gammay_gpu[x, y] + alpha2y_gpu[x, y] * eyy_gpu[x, y]
    end
    return
end


function DiffDxplus_kernel!(dA_gpu, A_gpu, w_gpu, dx, nx, ny, l)
    x = threadIdx().x + blockDim().x * (blockIdx().x - 1)
    y = threadIdx().y + blockDim().y * (blockIdx().y - 1)

    if x <= nx && y <= ny
        if x <= l
            sum = 0.0
            for k in 1:x
                sum = sum - w_gpu[k] * A_gpu[x - (k - 1), y]
            end
            for k in 1:l
                sum = sum + w_gpu[k] * A_gpu[x + k, y]
            end
            dA_gpu[x, y] = sum / dx
        elseif x >= l+1 && x <= nx-l
            sum = 0.0
            for k in 1:l
                sum = sum + w_gpu[k] * (-A_gpu[x - (k - 1), y] + A_gpu[x + k, y])
            end
            dA_gpu[x, y] = sum / dx
        else
            sum = 0.0
            for k in 1:l
                sum = sum - w_gpu[k] * A_gpu[x - (k - 1), y]
            end
            for k in 1:(nx - x)
                sum = sum + w_gpu[k] * A_gpu[x + k, y]
            end
            dA_gpu[x, y] = sum / dx
        end
    end
    return
end

function DiffDyplus_kernel!(dA_gpu, A_gpu, w_gpu, dx, nx, ny, l)
    x = threadIdx().x + blockDim().x * (blockIdx().x - 1)
    y = threadIdx().y + blockDim().y * (blockIdx().y - 1)

    if x <= nx
        if y >= 1 && y <= l
            sum = 0.0
            for k in 1:y
                sum -= w_gpu[k] * A_gpu[x, y - (k - 1)]
            end
            for k in 1:l
                sum += w_gpu[k] * A_gpu[x, y + k]
            end
            dA_gpu[x, y] = sum / dx

        elseif y >= l + 1 && y <= ny - l
            sum = 0.0
            for k in 1:l
                sum += w_gpu[k] * (-A_gpu[x, y - (k - 1)] + A_gpu[x, y + k])
            end
            dA_gpu[x, y] = sum / dx

        elseif y >= ny - l + 1 && y <= ny
            sum = 0.0
            for k in 1:l
                sum -= w_gpu[k] * A_gpu[x, y - (k - 1)]
            end
            for k in 1:(ny - y)
                sum += w_gpu[k] * A_gpu[x, y + k]
            end
            dA_gpu[x, y] = sum / dx
        end
    end

    return
end

function DiffDxminus_kernel!(dA_gpu, A_gpu, w_gpu, dx, nx, ny, l)
    x = threadIdx().x + blockDim().x * (blockIdx().x - 1)
    y = threadIdx().y + blockDim().y * (blockIdx().y - 1)

    if y <= ny
        if x >= 1 && x <= l
            sum = 0.0
            for k in 1:x
                if x - k >= 1
                    sum -= w_gpu[k] * A_gpu[x - k, y]
                end
            end
            for k in 1:l
                sum += w_gpu[k] * A_gpu[x + (k - 1), y]
            end
            dA_gpu[x, y] = sum / dx

        elseif x >= l + 1 && x <= nx - l
            sum = 0.0
            for k in 1:l
                sum += w_gpu[k] * (-A_gpu[x - k, y] + A_gpu[x + (k - 1), y])
            end
            dA_gpu[x, y] = sum / dx

        elseif x >= nx - l + 1 && x <= nx
            sum = 0.0
            for k in 1:l
                sum -= w_gpu[k] * A_gpu[x - k, y]
            end
            for k in 1:(nx - x)
                sum += w_gpu[k] * A_gpu[x + (k - 1), y]
            end
            dA_gpu[x, y] = sum / dx
        end
    end

    return
end

function DiffDyminus_kernel!(dA_gpu, A_gpu, w_gpu, dx, nx, ny, l)
    x = threadIdx().x + blockDim().x * (blockIdx().x - 1)
    y = threadIdx().y + blockDim().y * (blockIdx().y - 1)

    if x <= nx
        if y >= 1 && y <= l
            sum = 0.0
            for k in 1:y
                if y - k >= 1
                    sum -= w_gpu[k] * A_gpu[x, y - k]
                end
            end
            for k in 1:l
                sum += w_gpu[k] * A_gpu[x, y + (k - 1)]
            end
            dA_gpu[x, y] = sum / dx

        elseif y >= l + 1 && y <= ny - l
            sum = 0.0
            for k in 1:l
                sum += w_gpu[k] * (-A_gpu[x, y - k] + A_gpu[x, y + (k - 1)])
            end
            dA_gpu[x, y] = sum / dx

        elseif y >= ny - l + 1 && y <= ny
            sum = 0.0
            for k in 1:l
                sum -= w_gpu[k] * A_gpu[x, y - k]
            end
            for k in 1:(ny - y)
                sum += w_gpu[k] * A_gpu[x, y + (k - 1)]
            end
            dA_gpu[x, y] = sum / dx
        end
    end

    return
end

function update_pressure_kernel!(p_gpu::CuDeviceMatrix{Float64, 1}, rho_gpu::CuDeviceMatrix{Float64, 1}, src_val::Float64, Dt::Float64, Dx::Float64, sx::Int64, sy::Int64)
    p_gpu[sx, sy] += Dt * (src_val / (Dx * Dx * rho_gpu[sx, sy]))
    return nothing
end
